version: '3.8'

services:
  nginx:
    image: nginx:alpine
    ports:
      - "80:80"
    volumes:
      - ./nginx.conf:/etc/nginx/nginx.conf
    depends_on:
      - web
      - frontend
    networks:
      - app-network

  db:
    image: mysql:8.0
    environment:
      MYSQL_ROOT_PASSWORD: root_password
      MYSQL_DATABASE: instagram_clone_development 
      MYSQL_USER: app
      MYSQL_PASSWORD: app_password
    networks:
      - app-network
    healthcheck:
      test: ["CMD-SHELL", "mysqladmin ping -h localhost -u root --password=$$MYSQL_ROOT_PASSWORD"]
      interval: 5s
      timeout: 5s
      retries: 10   
    volumes:
      - mysql_data:/var/lib/mysql 

  web:
    build: 
      context: .
      dockerfile: Dockerfile
    image: artemromanenk0/instagram_clone:latest 
    restart: always
    volumes:
      - ./public:/app/public
    depends_on:
      db:
        condition: service_healthy
    environment:
      DATABASE_URL: mysql2://app:app_password@db/instagram_clone_development
      RAILS_ENV: development
    entrypoint: ["/bin/sh", "/app/docker-entrypoint.sh"]
    networks:
      - app-network

  frontend:
    build:
      context: ./frontend
      dockerfile: Dockerfile
    environment:
      NEXT_PUBLIC_API_URL: http://localhost/api # Через Nginx
      NODE_ENV: production
    networks:
      - app-network
    depends_on:
      - web

networks:
  app-network:  
    driver: bridge

volumes:
  mysql_data: